### PR TITLE
HPCC-13055 OSX support

### PR DESCRIPTION
**Pull request by richardkchapman**
Add resources using a generated assembler file with .incbin directives -
this should be more portable than binutils method which has proved
problematic on new architectures.

Read resources using mach-o support API.

Various changes in jcomp to support adding files to compile that are
not cpp files, and to support Apple toolchain.

Also add default:break; code to eclhelper to suppress warnings in generated
code, and add missing include files to allow getDiskUsage to work.

Signed-off-by: Richard Chapman rchapman@hpccsystems.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
@GordonSmith Please review

File: common/dllserver/thorplugin.cpp [hunk 1]
```diff
@@ -269,6 +276,35 @@ extern bool getResourceFromFile(const char *filename, MemoryBuffer &data, const
...
+    {
+        DBGLOG("Failed to load library %s: %d", filename, errno);
+        return false;
+    }
+    __uint64 size = stat_buf.st_size;
+    byte *start_addr = (byte *) mmap(0, size, PROT_READ, MAP_FILE | MAP_PRIVATE, fd, 0);
+    if (start_addr == MAP_FAILED)
+    {
+        DBGLOG("Failed to load library %s: %d", filename, errno);
+        return false;
```
Comment (GordonSmith) [inline common/dllserver/thorplugin.cpp]: Should you close fd here (try/finally might be better)?

File: common/dllserver/thorplugin.cpp [hunk 2]
```diff
@@ -269,6 +276,35 @@ extern bool getResourceFromFile(const char *filename, MemoryBuffer &data, const
...
+    }
+    // The first bytes are the Mach-O header
+    struct mach_header_64 *mh = (struct mach_header_64 *) start_addr;
+    if (mh->magic != MH_MAGIC_64)
+    {
+        DBGLOG("Failed to load library %s: Does not appear to be a Mach-O 64-bit binary", filename);
```
Comment (GordonSmith) [inline common/dllserver/thorplugin.cpp]: Should you close fd here (try/finally might be better)?

File: ecl/hqlcpp/hqlres.cpp [hunk 3]
```diff
@@ -463,17 +467,39 @@ void ResourceManager::flush(const char *filename, bool flushText, bool target64b
...
+        const char *type = s.type.str();
+        unsigned id = s.id;
+        VStringBuffer binfile("%s_%s_%u.bin", filename.str(), type, id);
+        fprintf(f, " .section __TEXT,%s_%u\n", type, id);
+        fprintf(f, " .global _%s_%u_txt_start\n", type, id);  // For some reason apple needs a leading underbar and linux does not
+        fprintf(f, "_%s_%u_txt_start:\n", type, id);
+        fprintf(f, " .incbin \"%s\"\n", binfile.str());
+        FILE *bin = fopen(binfile, "wb");
+        if (!bin)
+            throwError1(HQLERR_ResourceCreateFailed, binfile.str());
```
Comment (GordonSmith) [inline ecl/hqlcpp/hqlres.cpp]: Do care about closing "f" here?

File: rtl/include/eclhelper_base.hpp [hunk 4]
```diff
@@ -3259,6 +3488,8 @@ class CThorStreamedIteratorArg : public CThorArg, implements IHThorStreamedItera
         case TAIarg:
         case TAIstreamediteratorarg_1:
             return static_cast<IHThorStreamedIteratorArg *>(this);
+        default:
+            break;
```
Comment (GordonSmith) [inline rtl/include/eclhelper_base.hpp]: Just curious, are these to suppress compiler warnings?

File: system/jlib/jcomp.cpp [hunk 5]
```diff
@@ -666,11 +675,11 @@ bool CppCompiler::doLink()
     DWORD runcode = 0;
     if (verbose)
         PrintLog("%s", expanded.toCharArray());
-    StringBuffer logFile = StringBuffer(CORE_NAME).append("_link.log.tmp");
+    StringBuffer logFile = StringBuffer(coreName).append("_link.log.tmp");
     logFiles.append(logFile);
     bool ret = invoke_program(expanded.toCharArray(), runcode, true, logFile) && (runcode == 0);
-    linkFailed = !ret;
-    return ret;
+	linkFailed = !ret;
```
Comment (GordonSmith) [inline system/jlib/jcomp.cpp]: Indentation

**Comment by GordonSmith** [comment]
Just the possible leaked file handles, rest looks fine.

File: common/dllserver/thorplugin.cpp (hunk 2 above)
Comment (richardkchapman) [inline common/dllserver/thorplugin.cpp]: Yes, I should (and munmap). C++ support for finally is a little lacking though.

File: ecl/hqlcpp/hqlres.cpp (hunk 3 above)
Comment (richardkchapman) [inline ecl/hqlcpp/hqlres.cpp]: Yes, I probably should. Ideally I should have coded using a CFile that would have auto-destructed. In practice it's very unlikely to happen and even less likely to cause a problem though.

**Comment by richardkchapman** [comment]
Repushed to address concerns (and rebased to latest 5.2.0)

**Comment by GordonSmith** [comment]
Looks ok.

**Pull request merged**